db/range_del_aggregator_bench.cc

```diff
@@ -0,0 +1,197 @@
+  std::vector<std::string> keys, values;
+  for (const auto& range_del : range_dels) {
+    auto key_and_value = range_del.tombstone.Serialize();
+    keys.push_back(key_and_value.first.Encode().ToString());
+    values.push_back(key_and_value.second.ToString());
+  }
+  std::unique_ptr<test::VectorIterator> range_del_iter(
+      new test::VectorIterator(keys, values));
+  range_del_agg->AddTombstones(std::move(range_del_iter));
+}
+
+// convert long to a big-endian slice key
+static std::string Key(int64_t val) {
+  std::string little_endian_key;
+  std::string big_endian_key;
+  PutFixed64(&little_endian_key, val);
+  assert(little_endian_key.size() == sizeof(val));
+  big_endian_key.resize(sizeof(val));
+  for (size_t i = 0; i < sizeof(val); ++i) {
+    big_endian_key[i] = little_endian_key[sizeof(val) - 1 - i];
+  }
+  return big_endian_key;
+}
+
+}  // namespace rocksdb
+
+int main(int argc, char** argv) {
+  ParseCommandLineFlags(&argc, &argv, true);
+
+  Stats stats;
+  rocksdb::Random64 rnd(FLAGS_seed);
+  std::default_random_engine random_gen(FLAGS_seed);
+  std::normal_distribution<double> normal_dist(FLAGS_tombstone_width_mean,
+                                               FLAGS_tombstone_width_stddev);
+  std::vector<rocksdb::PersistentRangeTombstone> persistent_range_tombstones(
+      FLAGS_num_range_tombstones);
+  auto mode =
+      FLAGS_use_collapsed
+          ? rocksdb::RangeDelPositioningMode::kForwardTraversal
+          : rocksdb::RangeDelPositioningMode::kFullScan;
+
+  for (int i = 0; i < FLAGS_num_runs; i++) {
+    auto icmp = rocksdb::InternalKeyComparator(rocksdb::BytewiseComparator());
+    rocksdb::RangeDelAggregator range_del_agg(icmp, {} /* snapshots */,
+                                              FLAGS_use_collapsed);
+
+    for (int j = 0; j < FLAGS_num_range_tombstones; j++) {
+      uint64_t start = rnd.Uniform(FLAGS_tombstone_start_upper_bound);
+      uint64_t end = start + std::max(1.0, normal_dist(random_gen));
+      persistent_range_tombstones[j] = rocksdb::PersistentRangeTombstone(
+          rocksdb::Key(start), rocksdb::Key(end), j);
+    }
+
+    rocksdb::StopWatchNano stop_watch_add_tombstones(rocksdb::Env::Default(),
+                                                     true /* auto_start */);
+    rocksdb::AddTombstones(&range_del_agg, persistent_range_tombstones);
+    stats.time_add_tombstones += stop_watch_add_tombstones.ElapsedNanos();
+
+    uint64_t key = rnd.Uniform(FLAGS_should_delete_upper_bound);
+    rocksdb::ParsedInternalKey parsed_key;
+    std::string key_string = rocksdb::Key(key);
+    parsed_key.user_key = key_string;
+    parsed_key.sequence = FLAGS_num_range_tombstones / 2;
+    parsed_key.type = rocksdb::kTypeValue;
+
+    rocksdb::StopWatchNano stop_watch_should_delete(rocksdb::Env::Default(),
+                                                    true /* auto_start */);
+    range_del_agg.ShouldDelete(parsed_key, mode);
```

Comment:
Even when we're passing `kForwardTraversal`, it needs to do a binary search for the first call to `ShouldDelete`. We could make multiple calls on consecutive keys, such that the code paths for caching position and scanning forward in the range tombstone map can be exercised. This requires another gflag though.